psutil/arch/osx/smc.c

```diff
@@ -0,0 +1,228 @@
+    );
+    if (result != kIOReturnSuccess) {
+        return 1;
+    }
+
+    device = IOIteratorNext(iterator);
+    IOObjectRelease(iterator);
+    if (device == 0) {
+        return 1;
+    }
+
+    result = IOServiceOpen(device, mach_task_self(), 0, pconn);
+    IOObjectRelease(device);
+    if (result != kIOReturnSuccess) {
+        return 1;
+    }
+
+    return kIOReturnSuccess;
+}
+
+
+kern_return_t SMCClose(io_connect_t conn)
+{
+    return IOServiceClose(conn);
+}
+
+
+kern_return_t SMCCall(io_connect_t conn,
+    int index, SMCKeyData_t *inputStructure, SMCKeyData_t *outputStructure)
+{
+    size_t   structureInputSize;
+    size_t   structureOutputSize;
+
+    structureInputSize = sizeof(SMCKeyData_t);
+    structureOutputSize = sizeof(SMCKeyData_t);
+
+#if MAC_OS_X_VERSION_10_5
+    return IOConnectCallStructMethod(
+        conn, index,
+        inputStructure,
+        structureInputSize,
+        outputStructure,
+        &structureOutputSize);
+#else
+    return IOConnectMethodStructureIStructureO(
+        conn, index,
+        structureInputSize,
+        &structureOutputSize,
+        inputStructure,
+        outputStructure);
+#endif
+}
+
+
+kern_return_t SMCReadKey(io_connect_t conn, UInt32Char_t key, SMCVal_t *val) {
+    kern_return_t result;
+    SMCKeyData_t  inputStructure;
+    SMCKeyData_t  outputStructure;
+
+    memset(&inputStructure, 0, sizeof(SMCKeyData_t));
+    memset(&outputStructure, 0, sizeof(SMCKeyData_t));
+    memset(val, 0, sizeof(SMCVal_t));
+
+    inputStructure.key = _strtoul(key, 4, 16);
+    inputStructure.data8 = SMC_CMD_READ_KEYINFO;
+
+    result = SMCCall(
+        conn,
+        KERNEL_INDEX_SMC,
+        &inputStructure,
+        &outputStructure
+    );
+    if (result != kIOReturnSuccess)
+        return result;
+
+    val->dataSize = outputStructure.keyInfo.dataSize;
+    _ultostr(val->dataType, outputStructure.keyInfo.dataType);
+    inputStructure.keyInfo.dataSize = val->dataSize;
+    inputStructure.data8 = SMC_CMD_READ_BYTES;
+
+    result = SMCCall(
+        conn,
+        KERNEL_INDEX_SMC,
+        &inputStructure,
+        &outputStructure
+    );
+    if (result != kIOReturnSuccess)
+        return result;
+
+    memcpy(val->bytes, outputStructure.bytes, sizeof(outputStructure.bytes));
+
+    return kIOReturnSuccess;
+}
+
+
+double SMCGetTemperature(char *key) {
+    io_connect_t conn;
+    SMCVal_t val;
+    kern_return_t result;
+    int intValue;
+
+    result = SMCOpen(&conn);
+    if (result != kIOReturnSuccess) {
+        return 0.0;
+    }
+
+    result = SMCReadKey(conn, key, &val);
+    if (result == kIOReturnSuccess) {
+        // read succeeded - check returned value
+        if (val.dataSize > 0) {
+            if (strcmp(val.dataType, DATATYPE_SP78) == 0) {
+                // convert sp78 value to temperature
+                intValue = val.bytes[0] * 256 + (unsigned char)val.bytes[1];
+                SMCClose(conn);
+                return intValue / 256.0;
```

Comment:
Are you sure this math is correct? I bumped into this which is different: 
http://scarff.id.au/cgit/fancontrol.git/tree/smc.c#n325

Reply:
I am getting the exact readings I get in iStats

Reply:
Cool.